engine-nextgen/src/main/kotlin/graphql/nadel/enginekt/transform/hydration/NadelHydrationArgumentsBuilder.kt

```diff
@@ -0,0 +1,113 @@
+    ): Map<String, NormalizedInputValue> {
+        val sourceField = getSourceFieldDefinition(instruction)
+
+        return mapFrom(
+            instruction.arguments
+                .map {
+                    val argumentDef = sourceField.getArgument(it.name)
+                    it.name to makeInputValue(it, argumentDef, parentNode, hydrationField)
+                }
+        )
+    }
+
+    private fun makeInputValue(
+        argument: NadelHydrationArgument,
+        argumentDef: GraphQLArgument,
+        parentNode: JsonNode,
+        hydrationField: NormalizedField,
+    ): NormalizedInputValue {
+        return when (val valueSource = argument.valueSource) {
+            is ValueSource.ArgumentValue -> hydrationField.getNormalizedArgument(valueSource.argumentName)
+            is ValueSource.FieldValue -> makeInputValue(argumentDef) {
+                getFieldValue(valueSource, parentNode)
+            }
+        }
+    }
+
+    private fun getFieldValue(
+        valueSource: ValueSource.FieldValue,
+        parentNode: JsonNode,
+    ): AnyNormalizedInputValueValue {
+        val value = JsonNodeExtractor.getNodesAt(
+            rootNode = parentNode,
+            queryResultKeyPath = valueSource.pathToField,
+        ).emptyOrSingle()?.value
+
+        return when (value) {
+            is AnyList -> NormalizedInputValueValue.ListValue(value)
+            // TODO: I think this needs to be reconstructed with leaf values as graphql.language.Value
+            is AnyMap -> @Suppress("UNCHECKED_CAST") NormalizedInputValueValue.ObjectValue(value as JsonMap)
+            null -> NormalizedInputValueValue.AstValue(NullValue.newNullValue().build())
+            is Double -> NormalizedInputValueValue.AstValue(FloatValue.newFloatValue().value(
+                value.toBigDecimal(),
+            ).build())
+            is Float -> NormalizedInputValueValue.AstValue(FloatValue.newFloatValue().value(
+                value.toBigDecimal(),
+            ).build())
+            is Number -> NormalizedInputValueValue.AstValue(IntValue.newIntValue().value(
+                value.toLong().toBigInteger(),
+            ).build())
+            is String -> NormalizedInputValueValue.AstValue(StringValue.newStringValue().value(value).build())
+            is Boolean -> NormalizedInputValueValue.AstValue(BooleanValue.newBooleanValue().value(value).build())
+            else -> error("Unknown value type '${value.javaClass.name}'")
+        }
+    }
+
+    private inline fun makeInputValue(
+        argumentDef: GraphQLArgument,
+        valueFactory: () -> AnyNormalizedInputValueValue,
+    ): NormalizedInputValue {
+        return NormalizedInputValue(
+            GraphQLTypeUtil.simplePrint(argumentDef.type), // Type name
+            valueFactory().value,
+        )
+    }
+}
+
+private typealias AnyNormalizedInputValueValue = NormalizedInputValueValue<*>
+
+internal sealed class NormalizedInputValueValue<T> {
```

Comment:
Added this because I really didn't like how `NormalizedInputValue.value` was just an `Any` where you had to read the doc to find out what you needed.